nano-sync/src/utils.rs

```diff
@@ -0,0 +1,203 @@
+use algebra::mnt4_753::{
+    Fr as MNT4Fr, G1Projective as MNT4G1Projective, G2Projective as MNT4G2Projective,
+};
+use algebra::mnt6_753::{
+    Fr as MNT6Fr, G1Projective as MNT6G1Projective, G2Projective as MNT6G2Projective,
+};
+use algebra::{BigInteger768, FpParameters};
+use algebra_core::fields::Field;
+use algebra_core::ProjectiveCurve;
+use r1cs_std::prelude::*;
+use rand::{thread_rng, RngCore};
+
+use crate::compression::BeSerialize;
+
+/// Serializes a G1 point in the MNT4-753 curve.
+pub fn serialize_g1_mnt4(point: MNT4G1Projective) -> [u8; 95] {
+    let mut buffer = [0u8; 95];
+    BeSerialize::serialize(&point.into_affine(), &mut &mut buffer[..]).unwrap();
+    buffer
+}
+
+/// Serializes a G2 point in the MNT4-753 curve.
+pub fn serialize_g2_mnt4(point: MNT4G2Projective) -> [u8; 190] {
+    let mut buffer = [0u8; 190];
+    BeSerialize::serialize(&point.into_affine(), &mut &mut buffer[..]).unwrap();
+    buffer
+}
+
+/// Serializes a G1 point in the MNT6-753 curve.
+pub fn serialize_g1_mnt6(point: MNT6G1Projective) -> [u8; 95] {
+    let mut buffer = [0u8; 95];
+    BeSerialize::serialize(&point.into_affine(), &mut &mut buffer[..]).unwrap();
+    buffer
+}
+
+/// Serializes a G2 point in the MNT6-753 curve.
+pub fn serialize_g2_mnt6(point: MNT6G2Projective) -> [u8; 285] {
+    let mut buffer = [0u8; 285];
+    BeSerialize::serialize(&point.into_affine(), &mut &mut buffer[..]).unwrap();
+    buffer
+}
+
+/// Creates a random G1 point in the MNT4-753 curve.
+pub fn gen_rand_g1_mnt4() -> MNT4G1Projective {
+    let rng = &mut thread_rng();
```

Comment:
We should comment that these are not cryptographically "random" as thread_rng is not a CSPRNG.
As far as I can see, we're only using it for the parameter generation, right?

Reply:
Hmmm. But thread_rng is a CSPRNG.
[https://docs.rs/rand/0.6.1/rand/rngs/struct.ThreadRng.html](https://docs.rs/rand/0.6.1/rand/rngs/struct.ThreadRng.html)

Reply:
Interesting. Perhaps it was not always the case, at least I remember it was not. But ok, then this resolves it.